modules/programs/terraform.nix

```
@@ -0,0 +1,55 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.terraform;
+
+in {
+  meta.maintainers = [ maintainers.bryanhonof ];
+
+  options.programs.terraform = {
+
+    enable = mkEnableOption ''
+      Terraform is an open-source infrastructure as code software tool that
+      provides a consistent CLI workflow to manage hundreds of cloud services.
+      Terraform codifies cloud APIs into declarative configuration files.
+
+      https://www.terraform.io/
+    '';
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.terraform;
+      defaultText = "pkgs.terraform";
+      description = "The Terraform package to use";
+    };
+
+    providers = mkOption {
+      type = types.listOf types.package;
+      default = [ ];
+      description = "A list of Terraform providers.";
+    };
```

```suggestion
    providers = mkOption {
      type = hm.types.selectorFunction;
      default = providers: [ ];
      defaultText = literalExpression "providers: [ ]";
      example = literalExpression ''
        providers: [ providers.consul providers.dyn ];
      '';
      description = ''
        A list of Terraform providers.
        </para>
        <para>
        A list of available providers can be obtained with either of the following commands:
        <orderedlist numeration="arabic">
          <listitem>
            <para><command>nix-env -f '&lt;nixpkgs&gt;' -qaP -A terraform-providers</command></para>
          </listitem>
          <listitem>
            <para><command>nix search 'nixpkgs#terraform-providers'</command></para>
          </listitem>
        </orderedlist>
      '';
    };
```

Let's change this to a selector function instead.